examples/step-69/step-69.cc

```diff
@@ -0,0 +1,2218 @@
+
+    const unsigned int dofs_per_cell =
+      discretization->finite_element.dofs_per_cell;
+    const unsigned int n_q_points = discretization->quadrature.size();
+
+    /* This is the implementation of the scratch data required by WorkStream */
+    MeshWorker::ScratchData<dim> scratch_data(
+      discretization->mapping,
+      discretization->finite_element,
+      discretization->quadrature,
+      update_values | update_gradients | update_quadrature_points |
+        update_JxW_values,
+      discretization->face_quadrature,
+      update_normal_vectors | update_values | update_JxW_values);
+
+    {
+      TimerOutput::Scope t(
+        computing_timer,
+        "offline_data - assemble lumped mass matrix, and c_ij");
+
+      /* This is the implementation of the "worker" required by WorkStream */
+      const auto local_assemble_system = [&](const auto &cell,
+                                             auto &      scratch,
+                                             auto &      copy) {
+        auto &is_artificial             = copy.is_artificial;
+        auto &local_dof_indices         = copy.local_dof_indices;
+        auto &local_boundary_normal_map = copy.local_boundary_normal_map;
+        auto &cell_lumped_mass_matrix   = copy.cell_lumped_mass_matrix;
+        auto &cell_cij_matrix           = copy.cell_cij_matrix;
+
+        is_artificial = cell->is_artificial();
+        if (is_artificial)
+          return;
+
+        local_boundary_normal_map.clear();
+        cell_lumped_mass_matrix.reinit(dofs_per_cell, dofs_per_cell);
+        for (auto &matrix : cell_cij_matrix)
+          matrix.reinit(dofs_per_cell, dofs_per_cell);
+
+        const auto &fe_values = scratch.reinit(cell);
+
+        local_dof_indices.resize(dofs_per_cell);
+        cell->get_dof_indices(local_dof_indices);
+
+        std::transform(local_dof_indices.begin(),
+                       local_dof_indices.end(),
+                       local_dof_indices.begin(),
+                       [&](auto index) {
+                         return partitioner->global_to_local(index);
+                       });
+
+        /* We compute the local contributions for the  lumped mass
+         matrix entries m_i and and vectors c_ij */
+        for (unsigned int q_point = 0; q_point < n_q_points; ++q_point)
+          {
+            const auto JxW = fe_values.JxW(q_point);
+
+            for (unsigned int j = 0; j < dofs_per_cell; ++j)
+              {
+                const auto value_JxW = fe_values.shape_value(j, q_point) * JxW;
+                const auto grad_JxW  = fe_values.shape_grad(j, q_point) * JxW;
+
+                cell_lumped_mass_matrix(j, j) += value_JxW;
+
+                for (unsigned int i = 0; i < dofs_per_cell; ++i)
+                  {
+                    const auto value = fe_values.shape_value(i, q_point);
+                    for (unsigned int d = 0; d < dim; ++d)
+                      cell_cij_matrix[d](i, j) += (value * grad_JxW)[d];
```

Comment:
here you are grabbing the first dim components of grad_JxW and throw away the rest. Would it make sense to use an extractor? You also recompute value*grad_JxW...

Reply:
I am pretty confident that the optimizer can handle the `value * grad_JxW` temporary (while iterating) over all directions. The tricky part here is that an entry c_ij is actually a vector (due to the value - gradient pairing). I don't see how an extractor can help here.

Reply:
As a follow-up to @tjhei's comment, `(value * grad_JxW)[d]` takes dim times as many operations as `value * grad_JxW[d]` :-)